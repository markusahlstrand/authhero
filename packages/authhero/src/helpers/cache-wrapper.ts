import { DataAdapters, CacheAdapter } from "@authhero/adapter-interfaces";

/**
 * Configuration options for the cache wrapper
 */
export interface CacheOptions {
  /** Default TTL in seconds for cached items */
  defaultTtl: number;
  /** Custom TTLs for specific adapter methods (in seconds) */
  customTtls?: Record<string, number>;
  /** Methods to exclude from caching (e.g., ["users:update", "sessions:create"]) */
  excludeMethods?: string[];
  /** List of adapter entities to cache (if empty, all entities are cached) */
  cacheEntities?: string[];
  /** Cache adapter implementation */
  cache: CacheAdapter;
  /** Key prefix to ensure cache isolation (e.g., tenant ID) */
  keyPrefix?: string;
}

// Create a hash key from parameters for cache lookup
function createCacheKey(
  adapterName: string,
  methodName: string,
  args: any[],
  keyPrefix?: string,
): string {
  try {
    const baseKey = `${adapterName}:${methodName}:${JSON.stringify(args)}`;
    return keyPrefix ? `${keyPrefix}:${baseKey}` : baseKey;
  } catch (error) {
    // If arguments can't be stringified (e.g., contain circular references),
    // use a timestamp to ensure uniqueness but effectively disable caching
    const baseKey = `${adapterName}:${methodName}:${Date.now()}-${Math.random()}`;
    return keyPrefix ? `${keyPrefix}:${baseKey}` : baseKey;
  }
}

/**
 * Adds caching to data adapter methods using the new CacheAdapter interface
 * This wraps each method of the data adapter to cache its results based on input parameters
 */
export function addCaching(
  data: DataAdapters,
  options: CacheOptions,
): DataAdapters {
  const {
    cache,
    defaultTtl,
    customTtls = {},
    excludeMethods = [],
    cacheEntities = [],
    keyPrefix,
  } = options;

  const excludeMethodsSet = new Set(excludeMethods);
  const shouldCacheAllEntities = cacheEntities.length === 0;
  const entitiesToCache = new Set(cacheEntities);

  const wrappedAdapters: Record<string, any> = {};

  // Process each adapter
  for (const [adapterName, adapter] of Object.entries(data)) {
    // Check if this adapter entity should be cached
    const shouldCacheEntity =
      shouldCacheAllEntities || entitiesToCache.has(adapterName);

    const wrappedAdapter: Record<string, any> = {};

    // Process each method in the adapter
    for (const [methodName, method] of Object.entries(
      adapter as Record<string, any>,
    )) {
      if (typeof method === "function") {
        // If this adapter entity shouldn't be cached, just pass through the original method
        if (!shouldCacheEntity) {
          wrappedAdapter[methodName] = method;
          continue;
        }

        // Check if this method should be excluded from caching
        const methodKey = `${adapterName}:${methodName}`;
        if (excludeMethodsSet.has(methodKey)) {
          wrappedAdapter[methodName] = method;
          continue;
        }

        // Get TTL for this specific method or use default
        const ttl = customTtls[methodKey] ?? defaultTtl;

        // Determine if this is a write operation (generally should not be cached)
        const isWriteOperation = [
          "create",
          "update",
          "remove",
          "delete",
          "set",
          "used",
        ].includes(methodName);

        // Wrap the method with caching logic
        wrappedAdapter[methodName] = async (...args: any[]) => {
          // For write operations, we need to invalidate related cache entries
          if (isWriteOperation) {
            // Simple approach: clear all cache entries for this adapter
            // Note: Since our CacheAdapter doesn't have getKeysByPrefix, we can't do selective invalidation
            // In a production system, you might want to extend the CacheAdapter interface or
            // maintain a separate index of keys by adapter

            // For now, we'll just execute the write operation without cache invalidation
            // This means the cache might have stale data until TTL expires
            return await (method as any).apply(adapter, args);
          }

          // For read operations, try to get from cache first
          const cacheKey = createCacheKey(adapterName, methodName, args, keyPrefix);
          const cachedResult = await cache.get(cacheKey);

          if (cachedResult !== null) {
            return cachedResult;
          }

          // Cache miss, execute the original method
          const result = await (method as any).apply(adapter, args);

          // Cache the result - if TTL is 0, cache without expiration (useful for request-scoped caching)
          if (ttl >= 0) {
            await cache.set(cacheKey, result, ttl);
          }

          return result;
        };
      } else {
        // For non-function properties, just copy them as is
        wrappedAdapter[methodName] = method;
      }
    }

    wrappedAdapters[adapterName] = wrappedAdapter;
  }

  return wrappedAdapters as DataAdapters;
}
