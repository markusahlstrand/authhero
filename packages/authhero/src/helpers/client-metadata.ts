import { Client, LegacyClient } from "@authhero/adapter-interfaces";

/**
 * Converts a Client from the new format to LegacyClient format by extracting
 * legacy fields from client_metadata
 */
export function clientToLegacyClient(
  client: Client & { tenant: any; connections: any[] },
): LegacyClient {
  const metadata = client.client_metadata || {};

  return {
    ...client,
    // Extract legacy fields from client_metadata
    disable_sign_ups: metadata.disable_sign_ups === "true",
    email_validation: metadata.email_validation || "disabled",
  } as LegacyClient;
}

/**
 * Extracts disable_sign_ups from client metadata
 */
export function getDisableSignUps(client: Client): boolean {
  const metadata = client.client_metadata || {};
  return metadata.disable_sign_ups === "true";
}

/**
 * Extracts email_validation from client metadata
 */
export function getEmailValidation(client: Client): string {
  const metadata = client.client_metadata || {};
  return metadata.email_validation || "disabled";
}

/**
 * Updates client metadata with legacy field values
 */
export function setLegacyFieldsInMetadata(
  client: Partial<Client>,
  disable_sign_ups?: boolean,
  email_validation?: string,
): Partial<Client> {
  const metadata = { ...(client.client_metadata || {}) };

  if (disable_sign_ups !== undefined) {
    metadata.disable_sign_ups = disable_sign_ups.toString();
  }

  if (email_validation !== undefined) {
    metadata.email_validation = email_validation;
  }

  return {
    ...client,
    client_metadata: metadata,
  };
}
