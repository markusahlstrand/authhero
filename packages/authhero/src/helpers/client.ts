import { HTTPException } from "hono/http-exception";
import { Client, connectionSchema } from "authhero";
import { Bindings } from "../types";

export async function getClient(
  env: Bindings,
  clientId: string,
): Promise<Client> {
  const client = await env.data.clients.get(clientId);
  if (!client) {
    throw new HTTPException(403, { message: "Client not found" });
  }

  // Temporarily assume that this is the default client.
  const defaultClient = await env.data.clients.get("DEFAULT_CLIENT");

  const connections = defaultClient
    ? client.connections
        .map((connection) => {
          const defaultConnection = defaultClient?.connections?.find(
            (c) => c.name === connection.name,
          );

          const mergedConnection = connectionSchema.parse({
            ...defaultConnection,
            ...connection,
            options: {
              ...defaultConnection?.options,
              ...connection.options,
            },
          });

          return mergedConnection;
        })
        .filter((c) => c)
    : client.connections;

  return {
    ...client,
    web_origins: [
      ...(defaultClient?.web_origins || []),
      ...(client.web_origins || []),
      `${env.ISSUER}u/login`,
    ],
    allowed_logout_urls: [
      ...(defaultClient?.allowed_logout_urls || []),
      ...(client.allowed_logout_urls || []),
      env.ISSUER,
    ],
    callbacks: [
      ...(defaultClient?.callbacks || []),
      ...(client.callbacks || []),
      `${env.ISSUER}u/info`,
    ],
    connections,
    domains: [...(client.domains || []), ...(defaultClient?.domains || [])],
    tenant: {
      ...(defaultClient?.tenant || {}),
      ...client.tenant,
    },
  };
}
