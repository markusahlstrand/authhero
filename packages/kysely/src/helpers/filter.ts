import { Kysely, SelectQueryBuilder } from "kysely";
import { Database } from "../db";

export function luceneFilter<TB extends keyof Database>(
  db: Kysely<Database>,
  qb: SelectQueryBuilder<Database, TB, {}>,
  query: string,
  searchableColumns: string[],
) {
  const filters = query
    .split(/\s+/)
    // This handles queries that incorrectly are using a = instead of :
    .map((q) => q.replace(/^([^:]+)=/g, "$1:"))
    .map((filter) => {
      let isNegation = filter.startsWith("-");
      let key, value, isExistsQuery, operator;

      if (filter.startsWith("-_exists_:")) {
        key = filter.substring(10);
        isExistsQuery = true;
        isNegation = true;
      } else if (filter.startsWith("_exists_:")) {
        key = filter.substring(9);
        isExistsQuery = true;
        isNegation = false;
      } else if (filter.includes(":")) {
        isNegation = filter.startsWith("-");
        [key, value] = isNegation
          ? filter.substring(1).split(":")
          : filter.split(":");
        isExistsQuery = false;

        if (value.startsWith(">=")) {
          operator = ">=";
          value = value.substring(2);
        } else if (value.startsWith(">")) {
          operator = ">";
          value = value.substring(1);
        } else if (value.startsWith("<=")) {
          operator = "<=";
          value = value.substring(2);
        } else if (value.startsWith("<")) {
          operator = "<";
          value = value.substring(1);
        } else {
          operator = "=";
        }
      } else {
        key = null;
        value = filter;
        isExistsQuery = false;
      }

      return { key, value, isNegation, isExistsQuery, operator };
    });

  // Apply filters to the query builder
  filters.forEach(({ key, value, isNegation, isExistsQuery, operator }) => {
    if (key) {
      if (isExistsQuery) {
        if (isNegation) {
          qb = qb.where(key, "is", null);
        } else {
          qb = qb.where(key, "is not", null);
        }
      } else {
        if (isNegation) {
          switch (operator) {
            case ">":
              qb = qb.where(key, "<=", value);
              break;
            case ">=":
              qb = qb.where(key, "<", value);
              break;
            case "<":
              qb = qb.where(key, ">=", value);
              break;
            case "<=":
              qb = qb.where(key, ">", value);
              break;
            default:
              qb = qb.where(key, "!=", value);
          }
        } else {
          qb = qb.where(key, operator, value);
        }
      }
    } else {
      const { ref } = db.dynamic;
      // Generic single-word search across specified columns
      qb = qb.where((eb) =>
        eb.or(
          searchableColumns.map((col) => eb(ref(col), "like", `%${value}%`)),
        ),
      );
    }
  });

  return qb;
}
