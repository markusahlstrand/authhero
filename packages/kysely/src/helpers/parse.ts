/**
 * Parse a JSON string if defined, otherwise return the default value.
 * This is useful for parsing JSON strings from database storage.
 */
export function parseJsonIfDefined<T>(
  value: string | null | undefined,
  defaultValue: T,
): T {
  if (!value) {
    return defaultValue;
  }
  try {
    return JSON.parse(value) as T;
  } catch {
    return defaultValue;
  }
}

/**
 * Parse multiple JSON string properties of an object.
 * Properties that are undefined or null will use their default values.
 *
 * @param obj - The source object containing JSON string properties to parse
 * @param propertyDefaults - Map of property names to their default values
 * @param target - The target object to write parsed values to (defaults to a new object based on obj)
 * @returns The target object with parsed properties
 *
 * @example
 * const source = { flags: '{"enabled":true}', sessions: null, name: "Test" };
 * const result = parseJsonProperties(source, { flags: {}, sessions: {} });
 * // result = { flags: {enabled: true}, sessions: {}, name: "Test" }
 */
export function parseJsonProperties<T extends Record<string, any>>(
  obj: T,
  propertyDefaults: Partial<Record<keyof T, any>>,
  target: any = { ...obj },
): any {
  for (const prop in propertyDefaults) {
    if (prop in obj) {
      target[prop] = parseJsonIfDefined(
        obj[prop] as any,
        propertyDefaults[prop],
      );
    }
  }
  return target;
}
