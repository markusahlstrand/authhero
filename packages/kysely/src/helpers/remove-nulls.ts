export function removeNullProperties<T = any>(obj: unknown): T {
  // Return primitives and null/undefined as-is
  if (obj === null || obj === undefined || typeof obj !== "object") {
    return obj as T;
  }

  if (Array.isArray(obj)) {
    return obj.map((item) =>
      // Only recurse into objects; leave primitives untouched
      item !== null && typeof item === "object"
        ? removeNullProperties(item)
        : item,
    ) as T;
  }

  const clone: Record<string, any> = { ...(obj as Record<string, any>) };

  for (const key in clone) {
    const val = clone[key];
    if (val === null) {
      delete clone[key];
    } else if (val !== null && typeof val === "object") {
      if (Array.isArray(val)) {
        clone[key] = val.map((item: unknown) =>
          item !== null && typeof item === "object"
            ? removeNullProperties(item)
            : item,
        );
      } else {
        clone[key] = removeNullProperties(val);
      }
    }
  }

  return clone as T;
}
