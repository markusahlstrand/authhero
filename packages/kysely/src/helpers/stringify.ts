/**
 * Stringify a value to JSON if it's defined, otherwise return undefined.
 * This is useful for converting objects to JSON strings for database storage.
 */
export function stringifyIfDefined<T>(
  value: T | undefined,
): string | undefined {
  return value !== undefined ? JSON.stringify(value) : undefined;
}

/**
 * Stringify multiple properties of an object to JSON strings.
 * Only properties that are defined will be stringified.
 *
 * @param obj - The source object containing properties to stringify
 * @param properties - Array of property names to stringify
 * @param target - The target object to write stringified values to (defaults to a new object based on obj)
 * @returns The target object with stringified properties
 *
 * @example
 * const source = { flags: { enabled: true }, sessions: { timeout: 300 }, name: "Test" };
 * const result = stringifyProperties(source, ['flags', 'sessions']);
 * // result = { flags: '{"enabled":true}', sessions: '{"timeout":300}', name: "Test" }
 */
export function stringifyProperties<T extends Record<string, any>>(
  obj: T,
  properties: (keyof T)[],
  target: any = { ...obj },
): any {
  for (const prop of properties) {
    if (obj[prop] !== undefined) {
      target[prop] = JSON.stringify(obj[prop]);
    }
  }
  return target;
}

/**
 * Remove undefined and null properties from an object.
 * This keeps the SQL payload clean by only including defined values.
 */
export function removeUndefinedAndNull<T extends Record<string, any>>(
  obj: T,
): Partial<T> {
  const cleaned: any = {};
  for (const key in obj) {
    if (obj[key] !== undefined && obj[key] !== null) {
      cleaned[key] = obj[key];
    }
  }
  return cleaned;
}
