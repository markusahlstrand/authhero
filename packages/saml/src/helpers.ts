import { XMLBuilder, XMLParser } from "fast-xml-parser";
import { samlRequestSchema, SAMLResponseJSON } from "./types";
import { base64 } from "oslo/encoding";
import { nanoid } from "nanoid";
import { SamlSigner } from "./signer";

export interface SAMLMetadataParams {
  entityId: string;
  assertionConsumerServiceUrl: string;
  singleLogoutServiceUrl: string;
  certificates: string[];
}

export interface SAMLResponseParams {
  destination: string;
  inResponseTo: string;
  audience: string;
  issuer: string;
  email: string;
  notBefore?: string;
  notAfter?: string;
  responseId?: string;
  assertionId?: string;
  sessionNotOnOrAfter?: string;
  issueInstant?: string;
  sessionIndex: string;
  userId: string;
  signature?: {
    privateKeyPem: string;
    cert: string;
    kid: string;
  };
  encode?: boolean;
}

export async function inflateRaw(
  compressedData: Uint8Array,
): Promise<Uint8Array> {
  // TODO: this is not supported in Node.js, so we need to use a polyfill
  const ds = new DecompressionStream("deflate-raw");
  // Create a new Uint8Array to ensure we have an ArrayBuffer (not SharedArrayBuffer)
  const buffer = new Uint8Array(compressedData);
  const decompressedStream = new Blob([buffer]).stream().pipeThrough(ds);
  return new Uint8Array(await new Response(decompressedStream).arrayBuffer());
}

export async function inflateDecompress(input: string): Promise<string> {
  const decodedBytes = await base64.decode(input.replace(/ /g, "+"));

  try {
    // Try to decompress using pako
    const decompressed = await inflateRaw(decodedBytes);
    return new TextDecoder().decode(decompressed);
  } catch (error) {
    console.warn(
      "Decompression failed, assuming data is not compressed:",
      error,
    );
    // If decompression fails, assume the data is not compressed
    return new TextDecoder().decode(decodedBytes);
  }
}

export async function parseSamlRequestQuery(samlRequestQuery: string) {
  const samlRequesteXml = await inflateDecompress(samlRequestQuery);

  const parser = new XMLParser({
    attributeNamePrefix: "@_",
    alwaysCreateTextNode: true,
    ignoreAttributes: false,
  });
  const samlRequestJson = parser.parse(samlRequesteXml);

  return samlRequestSchema.parse(samlRequestJson);
}

export function createSamlMetadata(samlMetadataParams: SAMLMetadataParams) {
  // Create KeyDescriptor entries for each certificate
  const keyDescriptors = samlMetadataParams.certificates.map((cert) => ({
    "@_use": "signing",
    KeyInfo: {
      "@_xmlns": "http://www.w3.org/2000/09/xmldsig#",
      X509Data: {
        X509Certificate: cert,
      },
    },
  }));

  const samlMetadataJSON = {
    EntityDescriptor: {
      "@_entityID": samlMetadataParams.entityId,
      "@_xmlns": "urn:oasis:names:tc:SAML:2.0:metadata",
      IDPSSODescriptor: {
        "@_protocolSupportEnumeration":
          "urn:oasis:names:tc:SAML:2.0:protocol",
        KeyDescriptor: keyDescriptors,
        SingleLogoutService: [
          {
            "@_Binding": "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect",
            "@_Location": samlMetadataParams.singleLogoutServiceUrl,
          },
          {
            "@_Binding": "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST",
            "@_Location": samlMetadataParams.singleLogoutServiceUrl,
          },
        ],
        NameIDFormat: [
          "urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress",
          "urn:oasis:names:tc:SAML:2.0:nameid-format:persistent",
          "urn:oasis:names:tc:SAML:2.0:nameid-format:transient",
        ],
        SingleSignOnService: [
          {
            "@_Binding": "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect",
            "@_Location": samlMetadataParams.assertionConsumerServiceUrl,
          },
          {
            "@_Binding": "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST",
            "@_Location": samlMetadataParams.assertionConsumerServiceUrl,
          },
        ],
        Attribute: [
          {
            "@_Name":
              "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress",
            "@_NameFormat": "urn:oasis:names:tc:SAML:2.0:attrname-format:uri",
            "@_FriendlyName": "E-Mail Address",
            "@_xmlns": "urn:oasis:names:tc:SAML:2.0:assertion",
          },
          {
            "@_Name":
              "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname",
            "@_NameFormat": "urn:oasis:names:tc:SAML:2.0:attrname-format:uri",
            "@_FriendlyName": "Given Name",
            "@_xmlns": "urn:oasis:names:tc:SAML:2.0:assertion",
          },
          {
            "@_Name":
              "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name",
            "@_NameFormat": "urn:oasis:names:tc:SAML:2.0:attrname-format:uri",
            "@_FriendlyName": "Name",
            "@_xmlns": "urn:oasis:names:tc:SAML:2.0:assertion",
          },
          {
            "@_Name":
              "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname",
            "@_NameFormat": "urn:oasis:names:tc:SAML:2.0:attrname-format:uri",
            "@_FriendlyName": "Surname",
            "@_xmlns": "urn:oasis:names:tc:SAML:2.0:assertion",
          },
          {
            "@_Name":
              "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier",
            "@_NameFormat": "urn:oasis:names:tc:SAML:2.0:attrname-format:uri",
            "@_FriendlyName": "Name ID",
            "@_xmlns": "urn:oasis:names:tc:SAML:2.0:assertion",
          },
        ],
      },
    },
  };

  const builder = new XMLBuilder({
    ignoreAttributes: false,
    suppressEmptyNode: true,
    preserveOrder: false,
    format: true,
    attributeNamePrefix: "@_",
  });

  // Generate XML
  return builder.build(samlMetadataJSON);
}

export async function createSamlResponse(
  samlResponseParams: SAMLResponseParams,
  signer?: SamlSigner,
): Promise<string> {
  const notBefore = samlResponseParams.notBefore || new Date().toISOString();
  const notAfter =
    samlResponseParams.notAfter ||
    new Date(new Date(notBefore).getTime() + 10 * 60 * 1000).toISOString();
  const issueInstant = samlResponseParams.issueInstant || notBefore;
  const sessionNotOnOrAfter =
    samlResponseParams.sessionNotOnOrAfter || notAfter;
  const responseId = samlResponseParams.responseId || `_${nanoid()}`;
  const assertionId = samlResponseParams.assertionId || `_${nanoid()}`;

  const samlResponseJson: SAMLResponseJSON = [
    {
      "samlp:Response": [
        {
          "saml:Issuer": [{ "#text": samlResponseParams.issuer }],
        },
        {
          "samlp:Status": [
            {
              "samlp:StatusCode": [],
              ":@": { "@_Value": "urn:oasis:names:tc:SAML:2.0:status:Success" },
            },
          ],
        },
        {
          "saml:Assertion": [
            {
              "saml:Issuer": [
                {
                  "#text": samlResponseParams.issuer,
                },
              ],
            },
            {
              "saml:Subject": [
                {
                  "saml:NameID": [{ "#text": samlResponseParams.email }],
                  ":@": {
                    "@_Format":
                      "urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress",
                  },
                },
                {
                  "saml:SubjectConfirmation": [
                    {
                      "saml:SubjectConfirmationData": [],
                      ":@": {
                        "@_InResponseTo": samlResponseParams.inResponseTo,
                        "@_NotOnOrAfter": notAfter,
                        "@_Recipient": samlResponseParams.destination,
                      },
                    },
                  ],
                  ":@": { "@_Method": "urn:oasis:names:tc:SAML:2.0:cm:bearer" },
                },
              ],
            },
            {
              "saml:Conditions": [
                {
                  "saml:AudienceRestriction": [
                    {
                      "saml:Audience": [
                        {
                          "#text": samlResponseParams.audience,
                        },
                      ],
                    },
                  ],
                },
              ],
              ":@": {
                "@_NotBefore": notBefore,
                "@_NotOnOrAfter": notAfter,
              },
            },
            {
              "saml:AuthnStatement": [
                {
                  "saml:AuthnContext": [
                    {
                      "saml:AuthnContextClassRef": [
                        {
                          "#text":
                            "urn:oasis:names:tc:SAML:2.0:ac:classes:unspecified",
                        },
                      ],
                    },
                  ],
                },
              ],
              ":@": {
                "@_AuthnInstant": issueInstant,
                "@_SessionIndex": samlResponseParams.sessionIndex,
                "@_SessionNotOnOrAfter": sessionNotOnOrAfter,
              },
            },
            {
              "saml:AttributeStatement": [
                {
                  "saml:Attribute": [
                    {
                      ":@": {
                        "@_FriendlyName": "persistent",
                        "@_Name": "id",
                        "@_NameFormat":
                          "urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified",
                      },
                      "saml:AttributeValue": [
                        {
                          "#text": samlResponseParams.userId,
                          ":@": {
                            "@_xmlns:xs": "http://www.w3.org/2001/XMLSchema",
                            "@_xmlns:xsi":
                              "http://www.w3.org/2001/XMLSchema-instance",
                            "@_xsi:type": "xs:string",
                          },
                        },
                      ],
                    },
                  ],
                },
                {
                  "saml:Attribute": [
                    {
                      ":@": {
                        "@_Name": "email",
                        "@_NameFormat":
                          "urn:oasis:names:tc:SAML:2.0:attrname-format:basic",
                      },
                      "saml:AttributeValue": [
                        {
                          "#text": samlResponseParams.email,
                          ":@": {
                            "@_xmlns:xs": "http://www.w3.org/2001/XMLSchema",
                            "@_xmlns:xsi":
                              "http://www.w3.org/2001/XMLSchema-instance",
                            "@_xsi:type": "xs:string",
                          },
                        },
                      ],
                    },
                  ],
                },
                {
                  "saml:Attribute": [
                    {
                      ":@": {
                        "@_Name": "Role",
                        "@_NameFormat":
                          "urn:oasis:names:tc:SAML:2.0:attrname-format:basic",
                      },
                      "saml:AttributeValue": [
                        {
                          "#text": "manage-account",
                          ":@": {
                            "@_xmlns:xs": "http://www.w3.org/2001/XMLSchema",
                            "@_xmlns:xsi":
                              "http://www.w3.org/2001/XMLSchema-instance",
                            "@_xsi:type": "xs:string",
                          },
                        },
                      ],
                    },
                  ],
                },
                {
                  "saml:Attribute": [
                    {
                      ":@": {
                        "@_Name": "Role",
                        "@_NameFormat":
                          "urn:oasis:names:tc:SAML:2.0:attrname-format:basic",
                      },
                      "saml:AttributeValue": [
                        {
                          "#text": "default-roles-master",
                          ":@": {
                            "@_xmlns:xs": "http://www.w3.org/2001/XMLSchema",
                            "@_xmlns:xsi":
                              "http://www.w3.org/2001/XMLSchema-instance",
                            "@_xsi:type": "xs:string",
                          },
                        },
                      ],
                    },
                  ],
                },
                {
                  "saml:Attribute": [
                    {
                      ":@": {
                        "@_Name": "Role",
                        "@_NameFormat":
                          "urn:oasis:names:tc:SAML:2.0:attrname-format:basic",
                      },
                      "saml:AttributeValue": [
                        {
                          "#text": "offline_access",
                          ":@": {
                            "@_xmlns:xs": "http://www.w3.org/2001/XMLSchema",
                            "@_xmlns:xsi":
                              "http://www.w3.org/2001/XMLSchema-instance",
                            "@_xsi:type": "xs:string",
                          },
                        },
                      ],
                    },
                  ],
                },
                {
                  "saml:Attribute": [
                    {
                      ":@": {
                        "@_Name": "Role",
                        "@_NameFormat":
                          "urn:oasis:names:tc:SAML:2.0:attrname-format:basic",
                      },
                      "saml:AttributeValue": [
                        {
                          "#text": "view-profile",
                          ":@": {
                            "@_xmlns:xs": "http://www.w3.org/2001/XMLSchema",
                            "@_xmlns:xsi":
                              "http://www.w3.org/2001/XMLSchema-instance",
                            "@_xsi:type": "xs:string",
                          },
                        },
                      ],
                    },
                  ],
                },
                {
                  "saml:Attribute": [
                    {
                      ":@": {
                        "@_Name": "Role",
                        "@_NameFormat":
                          "urn:oasis:names:tc:SAML:2.0:attrname-format:basic",
                      },
                      "saml:AttributeValue": [
                        {
                          "#text": "uma_authorization",
                          ":@": {
                            "@_xmlns:xs": "http://www.w3.org/2001/XMLSchema",
                            "@_xmlns:xsi":
                              "http://www.w3.org/2001/XMLSchema-instance",
                            "@_xsi:type": "xs:string",
                          },
                        },
                      ],
                    },
                  ],
                },
                {
                  "saml:Attribute": [
                    {
                      ":@": {
                        "@_Name": "Role",
                        "@_NameFormat":
                          "urn:oasis:names:tc:SAML:2.0:attrname-format:basic",
                      },
                      "saml:AttributeValue": [
                        {
                          "#text": "manage-account-links",
                          ":@": {
                            "@_xmlns:xs": "http://www.w3.org/2001/XMLSchema",
                            "@_xmlns:xsi":
                              "http://www.w3.org/2001/XMLSchema-instance",
                            "@_xsi:type": "xs:string",
                          },
                        },
                      ],
                    },
                  ],
                },
              ],
            },
          ],
          ":@": {
            "@_xmlns": "urn:oasis:names:tc:SAML:2.0:assertion",
            "@_ID": assertionId,
            "@_IssueInstant": issueInstant,
            "@_Version": "2.0",
          },
        },
      ],
      ":@": {
        "@_xmlns:samlp": "urn:oasis:names:tc:SAML:2.0:protocol",
        "@_xmlns:saml": "urn:oasis:names:tc:SAML:2.0:assertion",
        "@_Destination": samlResponseParams.destination,
        "@_ID": responseId,
        "@_InResponseTo": samlResponseParams.inResponseTo,
        "@_IssueInstant": issueInstant,
        "@_Version": "2.0",
      },
    },
  ];

  const builder = new XMLBuilder({
    ignoreAttributes: false,
    suppressEmptyNode: true,
    preserveOrder: true,
  });

  // Generate XML
  let xmlContent = builder.build(samlResponseJson);

  if (samlResponseParams.signature && signer) {
    xmlContent = await signer.signSAML(
      xmlContent,
      samlResponseParams.signature.privateKeyPem,
      samlResponseParams.signature.cert,
    );
  }

  if (samlResponseParams.encode === false) {
    return xmlContent;
  }

  const encodedResponse = btoa(xmlContent);

  return encodedResponse;
}
