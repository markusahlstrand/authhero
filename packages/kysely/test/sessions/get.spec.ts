import { describe, expect, it } from "vitest";
import { getTestServer } from "../helpers/test-server";
import { nanoid } from "nanoid";

describe("sessions", () => {
  describe("get", () => {
    it.skip("should create and get a session", async () => {
      const db = await getTestServer();

      await db.tenants.create({
        id: "tenantId",
        name: "Test Tenant",
        audience: "https://example.com",
        sender_email: "login@example.com",
        sender_name: "SenderName",
      });

      // Add a client
      await db.applications.create("tenantId", {
        id: "clientId",
        client_secret: "clientSecret",
        name: "Test Client",
        callbacks: ["https://example.com/callback"],
        allowed_logout_urls: ["https://example.com/callback"],
        web_origins: ["https://example.com"],
        disable_sign_ups: false,
      });

      // Add a test user
      await db.users.create("tenantId", {
        email: "foo@example.com",
        email_verified: true,
        name: "Test User",
        nickname: "Test User",
        picture: "https://example.com/test.png",
        connection: "email",
        provider: "email",
        is_social: false,
        user_id: "email|userId",
      });

      const id = nanoid();

      const sessionData = {
        id,
        user_id: "email|userId",
        client_id: "clientId",
        used_at: "2021-01-01T00:00:00.000Z",
        device: {
          last_ip: "",
          initial_ip: "",
          last_user_agent: "",
          initial_user_agent: "",
          initial_asn: "",
          last_asn: "",
        },
        clients: [],
      };

      await db.sessions.create("tenantId", sessionData);

      const session = await db.sessions.get("tenantId", id);

      expect(session).matchSnapshot(sessionData);
    });
  });
});
