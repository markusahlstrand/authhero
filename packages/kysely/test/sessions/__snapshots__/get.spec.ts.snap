// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sessions > get > should create and get a session 1`] = `
{
  "authenticated_at": "2025-02-11T14:09:13.945Z",
  "client_id": "clientId",
  "clients": [],
  "created_at": "2025-02-11T14:09:13.945Z",
  "deleted_at": null,
  "device": {
    "initial_asn": "",
    "initial_ip": "",
    "initial_user_agent": "",
    "last_asn": "",
    "last_ip": "",
    "last_user_agent": "",
  },
  "expires_at": "2021-01-01T00:00:00.000Z",
  "id": "sessionId",
  "idle_expires_at": "2021-01-01T00:00:00.000Z",
  "last_interaction_at": "2025-02-11T14:09:13.945Z",
  "revoked_at": null,
  "session_id": null,
  "updated_at": "2025-02-11T14:09:13.945Z",
  "used_at": "2021-01-01T00:00:00.000Z",
  "user_id": "email|userId",
}
`;
